### PR TITLE
Extend multiwavelet testing - closes issue #223, depends on PR #259

### PR DESCRIPTION
**Pull request by stonecoldhughes**
Please review the [developer documentation](https://github.com/project-asgard/asgard/wiki/developing)
on the wiki of this project that contains help and requirements.

## Proposed changes

Describe what this PR changes and why.  If it closes an issue, link to it here
with [a supported keyword](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

## What type(s) of changes does this code introduce?
_Put an `x` in the boxes that apply._

- [ ] Bugfix
- [ ] New feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

### Does this introduce a breaking change?

- [ ] Yes
- [ ] No

## Checklist

_Put an x in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- [ ] this PR is up to date with current the current state of 'develop'
- [ ] code added or changed in the PR has been clang-formatted
- [ ] this PR adds tests to cover any new code, or to catch a bug that is being fixed
- [ ] documentation has been added (if appropriate)

### CODE REVIEW

**Comment by bmcdanie** [comment]
@stonecoldhughes please review CI failures.

also, name PRs for what they're doing (e.g. extend multiwavelet testing) with a little more detail in the description. finally, put closes #[issue_num] in the description so that the issue will be linked and automatically closed.

File: src/basis_tests.cpp [hunk 1]
```diff
@@ -60,26 +56,42 @@ void test_multiwavelet_gen(int const degree)
...
+    test_multiwavelet_gen<TestType>(degree, tol_factor);
   }
 
+  SECTION("Multiwavelet generation, degree = 2")
+  {
+    int const degree = 2;
+    test_multiwavelet_gen<TestType>(degree, tol_factor);
+  }
+
+  /* Captain! Add new tests here. How high should we go? */
```
Comment (bmcdanie) [inline src/basis_tests.cpp]: remove comment

File: src/batch_tests.cpp [hunk 2]
```diff
@@ -639,21 +639,24 @@ void test_batched_gemm(int const m, int const n, int const k, int const lda,
   auto const effect_c = [m, n](auto const c) {
     return fk::matrix<P, mem_type::const_view>(c, 0, m - 1, 0, n - 1);
   };
+
+  P const tol_factor = std::is_same<P, double>::value ? 1e-15 : 1e-7;
+
   for (int i = 0; i < num_batch; ++i)
   {
     // we use random vals for tests - multiply eps by this amount
```
Comment (bmcdanie) [inline src/batch_tests.cpp]: pls remove this old comment

File: src/boundary_conditions_tests.cpp [hunk 3]
```diff
@@ -6,7 +6,7 @@
 template<typename P>
 void test_boundary_condition_vector(PDE<P> &pde,
```
Comment (bmcdanie) [inline src/boundary_conditions_tests.cpp]: is this meant to be in this one?

Comment (bmcdanie) [reply src/boundary_conditions_tests.cpp]: ah I see this PR depends on your prior.

please include this in the PR description

File: src/lib_dispatch.hpp [hunk 4]
```diff
@@ -169,4 +169,106 @@ template<typename P>
 void getrs(char *trans, int *n, int *nrhs, P *A, int *lda, int *ipiv, P *b,
            int *ldb, int *info);
 
+extern template void
```
Comment (bmcdanie) [inline src/lib_dispatch.hpp]: we took all the extern template declarations out of all the headers - I re-removed them from the RMSE branch, so will need to re-remove these/rebase on that.

**Review by bmcdanie — APPROVED** [review]
ready to go - be sure to rebase on RMSE/develop so that we can get the extern template declarations out of lib dispatch.

**Review by bmcdanie — CHANGES_REQUESTED** [review]
please trim comment on line 84 basis tests. also rebase on RMSE or remove extern declarations from lib_dispatch that were re-introduced.

File: src/basis_tests.cpp (hunk 1 above)
Comment (stonecoldhughes) [reply src/basis_tests.cpp]: done

File: src/batch_tests.cpp (hunk 2 above)
Comment (stonecoldhughes) [reply src/batch_tests.cpp]: removed

File: src/boundary_conditions_tests.cpp (hunk 3 above)
Comment (stonecoldhughes) [reply src/boundary_conditions_tests.cpp]: done

File: src/lib_dispatch.hpp (hunk 4 above)
Comment (stonecoldhughes) [reply src/lib_dispatch.hpp]: done

**Review by bmcdanie — APPROVED** [review]
looks good

**Pull request merged**